src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -89,6 +89,12 @@ private Task<PingReply> DoSendPingCore(IPAddress address, byte[] buffer, int tim
                 if (!isAsync || error != Interop.IpHlpApi.ERROR_IO_PENDING)
                 {
                     Cleanup(isAsync);
+
+                    if (error != 0 && Enum.IsDefined(typeof(IPStatus), error))
```

Comment:
Why is `Enum.IsDefined(typeof(IPStatus)`? This should still work without this check, just output might not look nice when printed

Reply:
It's still possible for some non-`IPStatus` errors to be returned by `GetLastError()`. For those, `Win32Exception` is the appropriate path.

Reply:
The way this is written, if we ever decide to add an enum value to IPStatus, this will end up changing what exception type is thrown.  Why is Win32Exception the appropriate thing to do if the value isn't defined by IPStatus today?

Reply:
The Send* methods of the Ping class can return a PingReply or throw exceptions. Many of the exceptions are related to argument checking. The API can also throw a PingException which is a public type. I'm not sure why we throw a Win32Exception but perhaps it is embedded into a PingException that is the top-level exception thrown.

The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.networkinformation.pingexception?view=netframework-4.8) says when a PingException is thrown versus returning a PingReply:

>The Ping class throws this exception to indicate that while sending an Internet Control Message Protocol (ICMP) Echo request, a method called by the Ping class threw an unhandled exception. Applications should check the inner exception of a PingException object to identify the problem.
The Ping class does not throw this exception if the ICMP Echo request fails because of network, ICMP, or destination errors. For such errors, the Ping class returns a PingReply object with the relevant IPStatus value set in the Status property.

As a baseline, we should make sure we align with .NET Framework behaviors.

Reply:
And we should also make sure the behavior is consistent across the various sync and async versions of the Send* APIs as well as between Windows and non-Windows.

Reply:
> why not just always return the PingReply?

So this is a little imperfect because `IcmpSendEcho2` implements things ambiguously: in the same value (via `GetLastError()`), it can return either a standard win32 error for any exceptional cases (i.e. bad parameters, general system errors, etc.) or a non-standard code that maps to an `IPStatus`. Unfortunately the way this API is defined, there is no foolproof way to future-proof code to support unknown status codes, but we can get it good enough, I think.

Framework does a similar check, but using a range. If the Windows/iphlpapi team ever adds a new value outside of that range, it'll throw exceptions rather than return an unknown enum value:
https://github.com/microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/System/net/System/Net/NetworkInformation/ping.cs#L711-L714

> we should also make sure the behavior is consistent across the various sync and async versions

My plan here is to assume an exceptional case is any value below `IP_STATUS_BASE` (`11000`). This covers all the exceptional error states explicitly documented by `IcmpSendEcho2` and `Icmp6SendEcho2` and leaves the ceiling open for any new values. I'll apply this to both sync and async.

> as well as between Windows and non-Windows.

Based on CI passing my tests on Linux, it looks like at least timeouts are handled correctly. I will do a quick check of the code, though.

Reply:
Thanks, @scalablecory.